### ISSUE TITLE
Benchmark clients are all using the same thread (fixed in #1597)

### ISSUE THREAD

**Issue by mrzzzrm** [hyrise/hyrise#1383]
...especially, this means that they only use a single thread to translate and optimize queries and this could be a bottleneck. Join Order Benchmark queries can take non-negligible time to optimize.

I am not sure what the right course would be here? Surely we cannot reserve a full core for each client. Maybe translation and optimization should be Tasks as well?

(Also, I am not sure about calling `std::this_thread::sleep_for(std::chrono::milliseconds(10));` when there are "sufficient" queries running at the moment. This halts the "client thread" for a considerable time and could be problematic for these OLTP workloads, maybe? At least use condition variables to block and unblock the client thread?)

**Comment by mrks** [comment]
I suppose this was simply not an issue until now. Having translation and optimization in their own tasks sounds like a reasonable idea. Maybe check with @klauck on what the server does.

Regarding `std::this_thread::sleep_for(std::chrono::milliseconds(10));`: @lanice did you check if this makes a difference? What would happen if we would busy-wait?

**Comment by lanice** [comment]
Just ran the numbers, first with the current master, second with no sleep in the BenchmarkRunner, just busy-wait:
```
+-----------+------------------+-------+------------------+-------+--------+---------------------------------+
| Benchmark | prev. iter/s     | runs  | new iter/s       | runs  | change | p-value (significant if <0.001) |
+-----------+------------------+-------+------------------+-------+--------+---------------------------------+
| TPC-H 1   | 11.0653162002563 | 664   | 9.76666641235352 | 586   | -12%   |                          0.0022 |
| TPC-H 2   | 5.23291492462158 | 314   | 5.91666650772095 | 355   | +13%   |                          0.3415 |
| TPC-H 3   | 67.0804595947266 | 4025  | 32.6166648864746 | 1957  | -51%   |                          0.0000 |
| TPC-H 4   | 49.9120330810547 | 2995  | 36.7646675109863 | 2206  | -26%   |                          0.0000 |
| TPC-H 5   | 67.3563995361328 | 4042  | 39.4999961853027 | 2370  | -41%   |                          0.0000 |
| TPC-H 6   | 467.798095703125 | 10000 | 460.501281738281 | 10000 | -2%    |                          0.0000 |
| TPC-H 7   | 45.9794425964355 | 2759  | 33.7333297729492 | 2024  | -27%   |                          0.0000 |
| TPC-H 8   | 84.1795196533203 | 5051  | 55.3489723205566 | 3322  | -34%   |                          0.0000 |
| TPC-H 9   | 37.1819534301758 | 2231  | 29.3833312988281 | 1763  | -21%   |                          0.0000 |
| TPC-H 10  | 68.2256240844727 | 4094  | 55.9333305358887 | 3356  | -18%   |                          0.0000 |
| TPC-H 11  | 771.894226074219 | 10000 | 1094.5654296875  | 10000 | +42%   |                          0.0000 |
| TPC-H 12  | 238.213409423828 | 10002 | 229.398651123047 | 10000 | -4%    |                          0.0000 |
| TPC-H 13  | 22.2323207855225 | 1334  | 21.9999980926514 | 1320  | -1%    |                          0.4073 |
| TPC-H 14  | 627.636108398438 | 10012 | 590.421081542969 | 10000 | -6%    |                          0.0000 |
| TPC-H 16  | 41.7112464904785 | 2503  | 40.8499984741211 | 2451  | -2%    |                          0.0003 |
| TPC-H 18  | 12.0314884185791 | 722   | 5.54999971389771 | 333   | -54%   |                          0.0000 |
| TPC-H 19  | 326.552032470703 | 10003 | 300.863891601562 | 10000 | -8%    |                          0.0000 |
| TPC-H 22  | 280.836517333984 | 10002 | 261.021514892578 | 10000 | -7%    |                          0.0000 |
| average   |                  |       |                  |       | -14%   |                                 |
+-----------+------------------+-------+------------------+-------+--------+---------------------------------+
```

This is using all 224 cores on nemea, sf 1. But in the second case we basically can use only 223 cores, because one is busy-waiting for scheduling new tasks. Haven't benchmarked with less than 224 cores yet.

**Comment by mrzzzrm** [comment]
Condition variable seems the intuitive approach here, tried that yet? (Just asking, not implying you should do that (now))

**Comment by lanice** [comment]
But I agree, using tasks for translation etc. probably makes sense. When I initially started this, it was not an issue because I created the tasks completely beforehand. Now that they are being scheduled continuously during the benchmark, it is different.

**Comment by lanice** [comment]
> Condition variable seems the intuitive approach here, tried that yet?

Yes seems right, not tried yet though.

**Comment by toniSta** [comment]
I am currently implementing a solution for this issue. Therefore, I created a new task (called `PipelineExecutionTask`) that creates and executes the pipeline in a separate task. However, I am getting segfaults or `std::bad_alloc` errors within the SQLParser as soon as I simulate more than 1 client.  Full stack trace is located in [this gist](https://gist.github.com/toniSta/ea28c8821c807fffc611a85e12d1b810). I don't get rid of the feeling that the SQLParser is not fully thread-safe. Anyone who's familiar with that one?

The code is on the Branch `toni/pipeline_tasks`. I created a scenario in hyrisePlayground to reproduce the error. In addition, I created a method using the old way of creating and executing the pipeline. This one's not throwing an error.

**Comment by mrks** [comment]
https://stackoverflow.com/questions/1677142/flex-lex-yacc-bison-multithreaded-environment

**Comment by mrzzzrm** [comment]
The solution proposed in the SO post seems straight forward, do you feel comfortable implementing it in the parser, @toniSta? If not, I might be able to assist.

**Comment by toniSta** [comment]
I checked both files and noticed, that the necessary flags are already set (`%option reentrant` in `flex_lexer.l:34` and `%define api.pure full` (instead of `%pure-parser`) in `bison_parser.y:64`)

**Comment by mrzzzrm** [comment]
And `%define api.pure full` is supposed to be the same as `%pure-parser`?

**Comment by toniSta** [comment]
The SO post seems to be pretty old. Following the [documentation](https://www.gnu.org/software/bison/manual/html_node/Pure-Decl.html#Pure-Decl) `%define api.pure full` needs to be set.

**Comment by mrzzzrm** [comment]
Then maybe the parser uses `static` function members somewhere? Just guessing...

**Comment by mrks** [comment]
What happens if you throw tsan on it?

**Comment by toniSta** [comment]
> Then maybe the parser uses `static` function members somewhere?

We make the parser parse things via a static method call: `hsql::SQLParser::parse(sql, &parse_result);`. There are also some static functions in the bison/flex which seem to be automatically generated.

> What happens if you throw tsan on it?

`Segmentation fault (core dumped)` 
No warning, no error. Only asan detects a stack overflow.

**Comment by mrks** [comment]
Maybe `static std::stringstream strbuf;`? Please try to run something without quoted strings and see if the problem persists.

**Comment by toniSta** [comment]
Problem persists. But the line you mentioned might be interesting.

**Comment by toniSta** [comment]
Did not manage to fix this problem. Maybe someone else can have a look at this.

**Comment by mrzzzrm** [comment]
Here, this fixes it for me: https://github.com/hyrise/sql-parser/pull/110
